NETOBSERV-1216 PacketDrop enhancements

jpinsonneau opened the pull request:
Requires:
https://github.com/netobserv/flowlogs-pipeline/pull/470
https://github.com/netobserv/network-observability-operator/pull/401

This is a followup on [NETOBSERV-1064](https://issues.redhat.com/browse/NETOBSERV-1064) to improve Packet drop functionnality:
- moved ICMP type / code to `protocol`
- moved drop cause to `packets`
- improved descriptions with clickable popovers including link to documentations
- allow filter on drop cause / tcp state
- removed green color when when drop info are not provided
  - flows containing the info will keep green / red colors even if the feature is disabled

![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/11a16514-75b2-481f-a33c-f36b8de18961)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/6af07e97-3aa7-4212-b634-f5fd39639e4c)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/5e9367f8-ce42-4101-83bb-a974832fcc26)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/521c28bb-c3e7-4bb0-95e3-2d4684087b6d)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/f4e11f5a-58ae-44f5-ae97-1e2b524a46ef)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/aa897e02-47ec-41bb-a6fb-3a0a97242a8d)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/0f4cd238-1d90-465f-87a8-b629c77d2d6f)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/56ccc590-936f-44d8-aa82-05c4e17c50b4)

![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/4ae59434-6cfa-4849-a434-89ba312e6511)
![image](https://github.com/netobserv/network-observability-console-plugin/assets/91894519/119f46dd-8225-46da-bdff-fd464b1da501)

msherif1234 commented on web/locales/en/plugin__netobserv-plugin.json:
```diff
@@ -429,12 +429,21 @@
+  "A _LINUX_TCP_STATES_H TCP name like ESTABLISHED, SYN_SENT, SYN_RECV": "A _LINUX_TCP_STATES_H TCP name like ESTABLISHED, SYN_SENT, SYN_RECV",
+  "Packet drop latest cause": "Packet drop latest cause",
+  "Specify a single TCP drop cause.": "Specify a single TCP drop cause.",
+  "Specify a single TCP drop cause like:": "Specify a single TCP drop cause like:",
```
this isn't `TCP` specific here and below ?

jpinsonneau replied:
sorry my bad

jpinsonneau replied:
https://github.com/netobserv/network-observability-console-plugin/pull/362/commits/c29eb9f68ffddf4e92f26db9ab8ed802d2788be8

msherif1234 commented on web/src/utils/filter-definitions.ts:
```diff
@@ -448,6 +449,34 @@ export const getFilterDefinitions = (
         overlap: false
       },
+      {
+        id: 'pkt_drop_state',
```
well this filter be off when the dropped packet is for UDP as an example ?

jpinsonneau replied:
No specific mechanism is done here. The user can set 2 filters as `PktDropLatestState=TCP_XXX`, `Proto=UDP` with `match any` query option to get both dropped TCP & UDP flows

msherif1234 commented on web/src/utils/filter-options.ts:
```diff
@@ -6,6 +6,7 @@ import { splitResource, SplitStage } from '../model/resource';
 import { autoCompleteCache } from './autocomplete-cache';
 import { DNS_RCODES } from './dns';
 import { getPort, getService } from './port';
+import { DROP_CAUSES, DROP_STATES } from './tcp-drop';
```
pls rename the pkg to pkt-drop instead ?

jpinsonneau replied:
https://github.com/netobserv/network-observability-console-plugin/pull/362/commits/c29eb9f68ffddf4e92f26db9ab8ed802d2788be8

msherif1234 commented on web/src/utils/icmp.ts:
```diff
@@ -1,75 +1,100 @@
+    case 5: // ICMP_REDIRECT:
+      return ICMP_REDIRECT_CODES.find(v => v.value === c);
+    case 11: // ICMP_TIME_EXCEEDED:
+      return ICMP_TIME_EXCEEDED_CODES.find(v => v.value === c);
```
what about icmp6 codes and types they aren't the same as in icmp4 ?

jpinsonneau replied:
Correct:
https://www.iana.org/assignments/icmp-parameters/icmp-parameters.xhtml#icmp-parameters-types
https://www.iana.org/assignments/icmpv6-parameters/icmpv6-parameters.xhtml#icmpv6-parameters-2

I should implement both and switch according to for protocol field

```
1 | ICMP | Internet Control Message
vs 
58 | IPv6-ICMP | ICMP for IPv6
```

jpinsonneau replied:
https://github.com/netobserv/network-observability-console-plugin/pull/362/commits/eb1c5be0f8207647eaed45f581a6862131f6838f

msherif1234 commented on web/src/utils/tcp-drop.ts:
```diff
@@ -0,0 +1,224 @@
+  { value: 72, name: 'SKB_DROP_REASON_IPV6_NDISC_HOP_LIMIT', description: 'invalid hop limit' },
+  { value: 73, name: 'SKB_DROP_REASON_IPV6_NDISC_BAD_CODE', description: 'invalid NDISC icmp6 code' },
+  { value: 74, name: 'SKB_DROP_REASON_IPV6_NDISC_BAD_OPTIONS', description: 'invalid NDISC options' },
+  { value: 75, name: 'SKB_DROP_REASON_IPV6_NDISC_NS_OTHERHOST', description: 'NEIGHBOUR SOLICITATION' }
```
do we have to maintain this translation from number to string in two components ? if it makes more sense to have it here we probably removed from FLP and just keep it as a number to reduce future efforts adding new drops ?

jpinsonneau replied:
That's a tradeoff to decide. If we want the user to be able to type either `number` or `name` in the filters, we need these in console plugin.

msherif1234 replied:
I think we need to be able to filter on both as u have

jotak commented on web/src/components/netflow-record/record-field.tsx:
```diff
@@ -336,38 +396,34 @@ export const RecordField: React.FC<{
+            child = clickableContent(
+              flow.fields.PktDropLatestDropCause,
+              getDropCauseDescription(flow.fields.PktDropLatestDropCause as DROP_CAUSES_NAMES),
+              'https://elixir.bootlin.com/linux/latest/source/include/net/dropreason-core.h#L88'
```
is there a specific reason to use a URL from `elixir.bootlin.com` ? (I don't know about this site)
This one sounds more official / primary source: https://github.com/torvalds/linux/blob/master/include/net/dropreason-core.h ?

jotak commented on web/src/utils/filter-definitions.ts:
```diff
@@ -448,6 +449,34 @@ export const getFilterDefinitions = (
+        validate: rejectEmptyValue,
+        hint: t('Specify a single TCP state.'),
+        examples: `${t('Specify a single TCP state name like:')}
+        - ${t('A _LINUX_TCP_STATES_H number like 1, 2, 3')}
+        - ${t('A _LINUX_TCP_STATES_H TCP name like ESTABLISHED, SYN_SENT, SYN_RECV')}`,
```
Like you put a link to the drop reasons, we may have a link to these states definitions? https://github.com/torvalds/linux/blob/master/include/net/tcp_states.h

jotak commented:
looks good to me in general (two comments about the external links)
You could also add a couple of tests, for example in https://github.com/netobserv/network-observability-console-plugin/blob/main/web/src/components/filters/__tests__/autocomplete-filter.spec.tsx#L29-L51 we could have similar tests for the new autocompleted fields

jpinsonneau commented:
> You could also add a couple of tests, for example in https://github.com/netobserv/network-observability-console-plugin/blob/main/web/src/components/filters/__tests__/autocomplete-filter.spec.tsx#L29-L51 we could have similar tests for the new autocompleted fields

Currently the new filters are just rejecting empty values `validate: rejectEmptyValue,`
Do you think we should force the user to the defined list only ? This was made on purpose to allow any new value to be filtered on, even if we didn't defined these.

I tried to check for the popover suggestions but the `wrapper.find` function can't find it :disappointed: 
It seems that the `waitForComponentToPaint` is not enough to wait for the popover to show or maybe the popover doesn't even show in these tests.

msherif1234 commented:
/lgtm

Amoghrd commented:
/ok-to-test

jpinsonneau commented:
Rebased without changes

msherif1234 commented:
/lgtm

jpinsonneau commented:
Addressing feedback from [QE doc](https://docs.google.com/document/d/1reaaHZi5E803AESF_uneBiMuueRcaSnng1BL99evC8A):
- Filters still available when feature is disabled https://github.com/netobserv/network-observability-console-plugin/pull/362/commits/03d6282a681d3008018d7f6f8b1b93a2dcad0962

jpinsonneau commented:
While testing I noticed `Bytes` and `Packets` fields are not skipped when zero on FLP side.
That create inconsistency between flows and conversations. I've added a commit for that
https://github.com/netobserv/network-observability-console-plugin/pull/362/commits/06ed39be89303aa5d39f967bb9e5bedbf076a828

That makes this PR dependant of https://github.com/netobserv/flowlogs-pipeline/pull/470 one. Waiting for feedback about that approach
